File: java/src/main/java/org/rocksdb/ParsedEntryInfo.java
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).

package org.rocksdb;

import java.nio.ByteBuffer;

/**
 * Class to parse internal key to extract user key, entry type, sequence number.
 */
public class ParsedEntryInfo extends RocksObject {
  protected ParsedEntryInfo() {
    super(newParseEntryInstance());
  }

  @Override
  protected void disposeInternal(final long handle) {
    disposeInternalJni(handle);
  }

  /**
   * Returns the entryType of record in the sstFile.
   */
  public EntryType getEntryType() {
    assert (isOwningHandle());
    return EntryType.getEntryType(getEntryTypeJni(nativeHandle_));
  }

  /**
   * Returns the sequence number of the record in the sstFile.
   */
  public long getSequenceNumber() {
    return getSequenceNumberJni(nativeHandle_);
  }

  /**
   * Returns the user key of the record in the sstFile.
   */
  public byte[] getUserKey() {
    assert (isOwningHandle());
    return userKeyJni(nativeHandle_);
  }

  /**
   *
   * @param key Byte buffer to write the user key into.
   * @return length of the key.
   */
  public int userKey(final ByteBuffer key) {
    if (key == null) {
      throw new IllegalArgumentException("ByteBuffer parameters must not be null");
    }
    assert (isOwningHandle());
    final int result;
    if (key.isDirect()) {
      result = userKeyDirect(nativeHandle_, key, key.position(), key.remaining());
    } else {
      result = userKeyByteArray(
          nativeHandle_, key.array(), key.arrayOffset() + key.position(), key.remaining());
    }
    key.limit(Math.min(key.position() + result, key.limit()));
    return result;
  }

  /**
   * Parses internal key to get initialize the values in this class.
   * @param options options used while writing the sst file.
   * @param internalKey byte array containing the internal key.
   */
  public final void parseEntry(final Options options, final byte[] internalKey) {
    if (options == null || internalKey == null) {
      throw new IllegalArgumentException("ByteBuffer and options parameters must not be null");
    }
    assert (isOwningHandle());
    parseEntry(nativeHandle_, options.getNativeHandle(), internalKey, internalKey.length);
  }

  /**
   * Parses internal key to get initialize the values in this class.
   * @param options options used while writing the sst file.
   * @param internalKey ByteBuffer containing the internal key.
   */
  public final void parseEntry(final Options options, final ByteBuffer internalKey) {
    if (options == null || internalKey == null) {
      throw new IllegalArgumentException("ByteBuffer and options parameters must not be null");
    }
    assert (isOwningHandle());
    if (internalKey.isDirect()) {
      parseEntryDirect(nativeHandle_, options.getNativeHandle(), internalKey,
          internalKey.position(), internalKey.remaining());
    } else {
      parseEntryByteArray(nativeHandle_, options.getNativeHandle(), internalKey.array(),
          internalKey.arrayOffset() + internalKey.position(), internalKey.remaining());
    }
    internalKey.position(internalKey.limit());
  }

  private static native long newParseEntryInstance();

  private static native void parseEntry(
      final long handle, final long optionsHandle, final byte[] buffer, final int bufferLen);

  private static native void parseEntryDirect(
      final long handle, final long optionsHandle, final ByteBuffer buffer, final int bufferOffset,
      final int bufferLen);
  private static native void parseEntryByteArray(
      final long handle, final long optionsHandle, final byte[] buffer, final int bufferOffset, final int bufferLen);
  private static native int userKeyDirect(
      final long handle, final ByteBuffer target, final int bufferOffset, final int bufferLen);
  private static native int userKeyByteArray(
      final long handle, final byte[] target, final int bufferOffset, final int bufferLen);
  private static native byte[] userKeyJni(final long handle);
  private static native long getSequenceNumberJni(final long handle);
  private static native byte getEntryTypeJni(final long handle);
  private static native void disposeInternalJni(final long handle);
}

File: java/rocksjni/parsed_entry_info.cc
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).
//
// This file implements the "bridge" between Java and C++ and enables
// calling c++ ROCKSDB_NAMESPACE::Iterator methods from Java side.

#include <jni.h>
#include <stdlib.h>

#include "include/org_rocksdb_ParsedEntryInfo.h"
#include "rocksdb/options.h"
#include "rocksdb/types.h"
#include "rocksdb/utilities/types_util.h"
#include "rocksjni/cplusplus_to_java_convert.h"
#include "rocksjni/portal.h"

/*
 * Class:     org_rocksdb_ParsedEntryInfo
 * Method:    newParseEntryInstance
 * Signature: ()J
 */
jlong JNICALL Java_org_rocksdb_ParsedEntryInfo_newParseEntryInstance(
    JNIEnv * /*env*/, jclass /*cls*/) {
  ROCKSDB_NAMESPACE::ParsedEntryInfo *parsed_entry_info =
      new ROCKSDB_NAMESPACE::ParsedEntryInfo();
  return GET_CPLUSPLUS_POINTER(parsed_entry_info);
}

/*
 * Class:     org_rocksdb_ParsedEntryInfo
 * Method:    parseEntry
 * Signature: (JJ[BI)V
 */
void JNICALL Java_org_rocksdb_ParsedEntryInfo_parseEntry(
    JNIEnv *env, jclass /*cls*/, jlong handle, jlong options_handle,
    jbyteArray jtarget, jint len) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  auto *parsed_entry_info =
      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
  jbyte *target = env->GetByteArrayElements(jtarget, nullptr);
  if (target == nullptr) {
    ROCKSDB_NAMESPACE::OutOfMemoryErrorJni::ThrowNew(env,
           "Memory allocation failed in RocksDB JNI function");
    return;
  }
  ROCKSDB_NAMESPACE::Slice target_slice(reinterpret_cast<char *>(target), len);
  ROCKSDB_NAMESPACE::ParseEntry(target_slice, options->comparator,
                                parsed_entry_info);
}

/*
 * Class:     org_rocksdb_ParsedEntryInfo
 * Method:    parseEntryDirect
 * Signature: (JJLjava/nio/ByteBuffer;II)V
 */
void JNICALL Java_org_rocksdb_ParsedEntryInfo_parseEntryDirect(
    JNIEnv *env, jclass /*clz*/, jlong handle, jlong options_handle,
    jobject jbuffer, jint jbuffer_off, jint jbuffer_len) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  auto *parsed_entry_info =
      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
  auto parse = [&parsed_entry_info,
                &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
    ROCKSDB_NAMESPACE::ParseEntry(target_slice, options->comparator,
                                  parsed_entry_info);
  };
  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(parse, env, jbuffer, jbuffer_off,
                                          jbuffer_len);
}

/*
 * Class:     org_rocksdb_ParsedEntryInfo
 * Method:    parseEntryByteArray
 * Signature: (JJ[BII)V
 */
void JNICALL Java_org_rocksdb_ParsedEntryInfo_parseEntryByteArray(
    JNIEnv *env, jclass /*clz*/, jlong handle, jlong options_handle,
    jbyteArray jtarget, jint joff, jint jlen) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  auto *parsed_entry_info =
      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
  auto parse = [&parsed_entry_info,
                &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
    ROCKSDB_NAMESPACE::ParseEntry(target_slice, options->comparator,
                                  parsed_entry_info);
  };
  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(parse, env, jtarget, joff, jlen);
}

/*
 * Class:     org_rocksdb_ParsedEntryInfo
 * Method:    userKeyDirect
 * Signature: (JLjava/nio/ByteBuffer;II)I
 */
jint JNICALL Java_org_rocksdb_ParsedEntryInfo_userKeyDirect(
    JNIEnv *env, jclass /*clz*/, jlong handle, jobject jtarget, jint joffset,
    jint jlen) {
  auto *parsed_entry_info =
      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
  ROCKSDB_NAMESPACE::Slice key_slice = parsed_entry_info->user_key;
  return ROCKSDB_NAMESPACE::JniUtil::copyToDirect(env, key_slice, jtarget,
                                                  joffset, jlen);
}

/*
 * Class:     org_rocksdb_ParsedEntryInfo
 * Method:    userKeyByteArray
 * Signature: (J[BII)I
 */
jint JNICALL Java_org_rocksdb_ParsedEntryInfo_userKeyByteArray(
    JNIEnv *env, jclass /*clz*/, jlong handle, jbyteArray jtarget, jint joffset,
    jint jlen) {
  auto *parsed_entry_info =
      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
  ROCKSDB_NAMESPACE::Slice key_slice = parsed_entry_info->user_key;
  return ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(env, key_slice, jtarget,
                                                     joffset, jlen);
}

/*
 * Class:     org_rocksdb_ParsedEntryInfo
 * Method:    userKeyJni
 * Signature: (J)[B
 */
jbyteArray JNICALL Java_org_rocksdb_ParsedEntryInfo_userKeyJni(JNIEnv *env,
                                                               jclass /*clz*/,
                                                               jlong handle) {
  auto *parsed_entry_info =
      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
  ROCKSDB_NAMESPACE::Slice key_slice = parsed_entry_info->user_key;
  jbyteArray jkey = env->NewByteArray(static_cast<jsize>(key_slice.size()));
  if (jkey == nullptr) {
    // exception thrown: OutOfMemoryError
    ROCKSDB_NAMESPACE::OutOfMemoryErrorJni::ThrowNew(env, "Memory allocation failed in RocksDB JNI function");
    return nullptr;
  }
  ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(
      env, key_slice, jkey, 0, static_cast<jint>(key_slice.size()));
  return jkey;
}

/*
 * Class:     org_rocksdb_ParsedEntryInfo
 * Method:    getSequenceNumberJni
 * Signature: (J)J
 */
jlong JNICALL Java_org_rocksdb_ParsedEntryInfo_getSequenceNumberJni(
    JNIEnv * /*env*/, jclass /*clz*/, jlong handle) {
  auto *parsed_entry_info =
      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
  uint64_t sequence_number = parsed_entry_info->sequence;
  return static_cast<jlong>(sequence_number);
}

/*
 * Class:     org_rocksdb_ParsedEntryInfo
 * Method:    getValueTypeJni
 * Signature: (J)B
 */
jbyte JNICALL Java_org_rocksdb_ParsedEntryInfo_getEntryTypeJni(JNIEnv * /*env*/,
                                                               jclass /*clz*/,
                                                               jlong handle) {
  auto *parsed_entry_info =
      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
  ROCKSDB_NAMESPACE::EntryType type = parsed_entry_info->type;
  return ROCKSDB_NAMESPACE::EntryTypeJni::toJavaEntryType(type);
}

/*
 * Class:     org_rocksdb_ParsedEntryInfo
 * Method:    disposeInternalJni
 * Signature: (J)V
 */
void JNICALL Java_org_rocksdb_ParsedEntryInfo_disposeInternalJni(
    JNIEnv * /*env*/, jclass /*clz*/, jlong handle) {
  auto *parsed_entry_info =
      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
  assert(parsed_entry_info != nullptr);
  delete parsed_entry_info;
}
